rpcs3/rpcs3qt/rsx_debugger.cpp

```diff
@@ -543,80 +641,544 @@ void rsx_debugger::GetMemory() const
+
+		pnl->showImage(QImage(pnl->cache.data(), width, height, format));
 	}
 
-	// Draw Texture
-	//if (!render->textures[m_cur_texture].enabled())
-	//	return;
+	// One iteration for depth, second for stencil
+	for (u32 buffer_it = 0; buffer_it < 2; buffer_it++)
+	{
+		const u32 rsx_buffer_addr = render->get_zeta_surface_address();
 
-	//u32 offset = render->textures[m_cur_texture].offset();
+		const u32 width  = rsx::method_registers.surface_clip_width();
+		const u32 height = rsx::method_registers.surface_clip_height();
+		const u32 pitch = rsx::method_registers.surface_z_pitch();
 
-	//if(!offset)
-	//	return;
+		u32 bpp = 4;
+		QImage::Format format = QImage::Format_Grayscale16;
+		const auto fmt = rsx::method_registers.surface_depth_fmt();
+		bool has_stencil = false;
+		bool is_stencil = buffer_it == 1;
+		bool is_float = false;
 
-	//u8 location = render->textures[m_cur_texture].location();
+		switch (fmt)
+		{
+		case rsx::surface_depth_format2::z16_float:
+		{
+			is_float = true;
+			[[fallthrough]];
+		}
+		case rsx::surface_depth_format2::z16_uint:
+		{
+			if (is_stencil)
+			{
+				continue;
+			}
 
-	//if(location > 1)
-	//	return;
+			bpp = 2;
+			break;
+		}
+		case rsx::surface_depth_format2::z24s8_float:
+		{
+			is_float = true;
+			[[fallthrough]];
+		}
+		case rsx::surface_depth_format2::z24s8_uint:
+		{
+			format = QImage::Format_RGB16;
+			has_stencil = true;
+			break;
+		}
+		default:
+		{
+			ensure(false);
+			break;
+		}
+		}
 
-	//u32 TexBuffer_addr = rsx::get_address(offset, location);
+		const u32 mem_size = pitch * (height - 1) + width * bpp;
 
-	//if(!vm::check_addr(TexBuffer_addr))
-	//	return;
+		Buffer* pnl{};
+		switch (buffer_it)
+		{
+		case 0:  pnl = m_buffer_depth; break;
+		default: pnl = m_buffer_stencil; break;
+		}
 
-	//unsigned char* TexBuffer = vm::get_super_ptr<u8>(TexBuffer_addr);
+		if (!height || !mem_size || !vm::check_addr(rsx_buffer_addr, vm::page_readable, mem_size))
+		{
+			if (render->is_paused())
+			{
+				pnl->showImage(QImage());
+			}
 
-	//const u32 width  = render->textures[m_cur_texture].width();
-	//const u32 height = render->textures[m_cur_texture].height();
-	//unsigned char* buffer = (unsigned char*)malloc(width * height * 3);
-	//std::memcpy(buffer, vm::base(TexBuffer_addr), width * height * 3);
+			continue;
+		}
+
+		// Touch RSX memory to potentially flush GPU memory (must occur in named_thread)
+		named_thread("RSX Buffer Touch", [&]()
+		{
+			for (u32 i = rsx_buffer_addr & -4096; i < rsx_buffer_addr + mem_size; i += 4096)
+			{
+				static_cast<void>(vm::_ref<atomic_t<u8>>(i).load());
+			}
+		});
+
+		const auto rsx_buffer = vm::get_super_ptr<const u8>(rsx_buffer_addr);
+		pnl->cache.resize(std::max<usz>(pnl->cache.size(), width * height * 4));
+		const auto buffer = pnl->cache.data();
+
+		if (is_stencil)
+		{
+			format = QImage::Format_Grayscale8;
+
+			for (u32 i = 0; i < mem_size; i += (pitch ? pitch : ~i))
```

Comment:
Iterating over pitch is really bad for readability and maintainability.
Just iterate over y and x everywhere